File: src/bridge/commands/semantic.ts
import { invoke } from '@tauri-apps/api/core'

import { requireNative } from '../native.ts'
import { authorizeSensitiveOperation } from './authorization.ts'

/** Semantic hit overlaying BM25 keyword results by note path. */
export interface SemanticSearchHit {
  note_path: string
  score: number
}

/** Thrown when semantic search is not configured for the vault. */
export class SemanticUnavailableError extends Error {
  constructor(context: string) {
    super(`${context}: semantic search is not configured for this vault`)
    this.name = 'SemanticUnavailableError'
  }
}

function parseSemanticPayload<T>(payload: string, context: string): T {
  const parsed = JSON.parse(payload) as Record<string, unknown>
  if (parsed.available === false) {
    // The vault has no `semantic` section; callers fall back to keyword search.
    throw new SemanticUnavailableError(context)
  }
  return parsed as T
}

export async function semanticSearch(query: string, limit = 25): Promise<SemanticSearchHit[]> {
  requireNative()
  const authorizationToken = await authorizeSensitiveOperation('ai_network_request', query)
  const payload = await invoke<string>('semantic_search', {
    query,
    limit,
    authorizationToken,
  })
  return parseSemanticPayload<SemanticSearchHit[]>(payload, 'semantic search')
}

export interface SemanticSyncReport {
  total_notes: number
  embedded: number
  unchanged: number
  removed: number
}

export async function semanticSync(): Promise<SemanticSyncReport> {
  requireNative()
  const authorizationToken = await authorizeSensitiveOperation(
    'ai_network_request',
    'Re-embed changed notes for semantic search',
  )
  const payload = await invoke<string>('semantic_sync', { authorizationToken })
  return parseSemanticPayload<SemanticSyncReport>(payload, 'semantic sync')
}

File: apps/desktop/src-tauri/src/commands/semantic.rs
//! Semantic (embedding) search commands. The daemon owns the engine;
//! these wrappers authorize the AI-network call, attach the OpenAI key
//! from the OS keychain (Ollama needs none), and surface the graceful
//! "unavailable" payload so the UI can fall back to keyword search. The
//! embeddings crate itself stays daemon-side by contract.

use scriptor_ipc::{RpcMethod, RpcPayload};
use scriptor_system_bridge::keychain_get;

use crate::authorization::{SensitiveOperation, require_sensitive_operation};
use crate::commands::daemon::daemon_rpc;
use crate::state::AppState;

/// Keychain account holding the user-supplied OpenAI key for semantic
/// embeddings. Ollama users never need this.
const SEMANTIC_OPENAI_KEYCHAIN_ACCOUNT: &str = "semantic.openai.key";

fn semantic_payload(payload: RpcPayload) -> Result<String, String> {
    match payload {
        RpcPayload::Json { json } => Ok(json),
        _ => Err("unexpected daemon semantic response".into()),
    }
}

/// Semantic search: returns `{"available":false}` when the vault has no
/// `semantic` section, letting callers fall back to keyword search.
#[tauri::command]
pub fn semantic_search(
    state: tauri::State<AppState>,
    query: String,
    limit: Option<u32>,
    authorization_token: String,
) -> Result<String, String> {
    require_sensitive_operation(
        &state,
        &authorization_token,
        SensitiveOperation::AiNetworkRequest,
        None,
    )?;
    let api_key =
        keychain_get(SEMANTIC_OPENAI_KEYCHAIN_ACCOUNT).map_err(|error| error.to_string())?;
    daemon_rpc(RpcMethod::EmbeddingsSearch {
        query,
        limit: limit.unwrap_or(25),
        api_key,
    })
    .and_then(semantic_payload)
}

/// Re-embed changed notes. Same opt-in contract as search.
#[tauri::command]
pub fn semantic_sync(
    state: tauri::State<AppState>,
    authorization_token: String,
) -> Result<String, String> {
    require_sensitive_operation(
        &state,
        &authorization_token,
        SensitiveOperation::AiNetworkRequest,
        None,
    )?;
    let api_key =
        keychain_get(SEMANTIC_OPENAI_KEYCHAIN_ACCOUNT).map_err(|error| error.to_string())?;
    daemon_rpc(RpcMethod::EmbeddingsSync { api_key }).and_then(semantic_payload)
}
